Zefi improvement

tbursztyka opened the pull request:
Got some time to rework this old base of code I made.

Try the PR on tests/arch/x86/info  on EHL with OEM UEFI BIOS for instance, ACPI is working there now. Thus PCIe MMIO, thus SMP, thus VT-D etc...

Commit:
zefi: Improve generic EFI header

This will prove to be useful to get a better EFI support.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
zefi: Expose EFI configuration lookup function

This will be useful to get various information such as ACPI table
pointer etc...

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
arch/x86: Get ACPI RSDP from EFI

EFI may have provide that pointer alread, so let's get it first.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
arch/x86: Prepare EFI support

As for Multiboot, let prep_c be aware of EFI boot.
In the futur, EFI will pass an argument to it.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
arch/x86: Expose multiboot init function even when disabled

Just a dummy function will do.

When enabled, the code does not need the #ifdef as cmake is handling
this properly already. This was also the wrong CONFIG_ used there
anyway.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
arch/x86: Expand cpu boot argument

In order to mitigate at runtime whether it booted on multiboot or EFI,
let's introduce a dedicated x86 cpu argument structure which holds the
type and the actual pointer delivered by the method (multiboot_info, or
efi_system_table)

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
zefi: Adding status code to header

This will be usefull when calling EFI functions.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
zefi: Add an EFI boot argument passing ACPI RSDP info

If such table pointer is present with EFI system table, this will speed
up ACPI initialization later on.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

andyross commented on arch/x86/core/intel64/cpu.c:
```diff
@@ -90,6 +90,15 @@ struct x86_tss64 tss3 = {
 #endif
 
+/* We must initialize with something OTHER THAN 0, or else it will land into
+ * .bss: in this case, though locore.S initalizes it relevantly, all will be
```
Might consider using an explicit section directive here instead of relying on the side effect of a noop initialization.

tbursztyka replied:
That would indeed be clearer.

andyross commented on arch/x86/zefi/zefi.c:
```diff
@@ -123,7 +123,8 @@ uintptr_t __abi efi_entry(void *img_handle, struct efi_system_table *sys_tab)
 
-	__asm__ volatile("cli; jmp *%0" :: "r"(code));
+	__asm__ volatile("cli; movq %0, %%rbx; jmp *%1"
+			 :: "r"(NULL), "r"(code) : );
```
Property, this needs a clobber declaration for RBX.   Obviously it's not a problem in practice because we're about to return (and the volatile asm block means nothing will reorder after us).  But still nice to be futureproof.

andyross commented on arch/x86/core/acpi.c:
```diff
@@ -27,21 +28,34 @@ static void find_rsdp(void)
 		return;
 	}
 
+	/* Let's first get it from EFI, if enabled */
```
Does this break test coverage for the legacy ACPI methods?  I'm pretty sure all our hardware platforms are EFI, and qemu doesn't do ACPI as we set it up, right?

tbursztyka replied:
Qemu can provide a basic support of ACPI, see tests/arch/x86/info (both qemu_x86 and qemu_x86_64). So as long as CONFIG_X86_EFI is not enabled on qemu (it can be, but it's a bit of setup work, see https://docs.zephyrproject.org/latest/boards/x86/qemu_x86/doc/index.html#programming-and-debugging ), the legacy ACPI will be covered.

andyross reviewed (dismissed):
Some nitpicks but nothing looks wrong.  Nice to see the EFI stuff getting some attention (even if moving in the direction of EFI support in the runtime OS is a **really** deep hole to start digging...)

tejlmand commented on arch/x86/Kconfig:
```diff
@@ -479,6 +479,14 @@ config X86_KPTI
 	  increase for additional page tables and trampoline stacks.
 
+config X86_EFI
+	bool "Enable EFI support"
```
As per the guidelines, please remove `Enable` and `support`:
https://docs.zephyrproject.org/latest/guides/build/kconfig/tips.html#prompt-strings

> For a Kconfig symbol that enables a driver/subsystem FOO, consider having just “Foo” as the prompt, instead of “Enable Foo support” or the like.

Maybe just write:
```suggestion
	bool "EFI"
```

tbursztyka replied:
Ok, did not know about such rule. Sounds a bit blunt in a menu to just read "EFI" but ok

tejlmand replied:
> Sounds a bit blunt in a menu to just read "EFI" but ok

You are always allowed to rephrase to a more descriptive title.
Just avoid the `Enable` and `support` as this adds no informational value.

Perhaps a title like:
```
	bool "EFI boot"
```
Already now, you'll notice that `boot` adds informational value to users just looking at the menu, as `boot` ensures that users couple `EFI` word to correct context (generally EFI is foreign in embedded systems).

There are many ways a title can be improved and provide extra informational value when removing the words `enable` and `support`. Better use the space for words of informational value, than to use those two words :wink:

tejlmand commented on arch/x86/zefi/zefi.py:
```diff
@@ -107,8 +107,8 @@ def build_elf(elf_file):
     #  + UEFI API assumes 16-bit wchar_t
     cmd = [args.compiler, "-shared", "-Wall", "-Werror", "-I.",
-        "-fno-stack-protector", "-fpic", "-mno-red-zone", "-fshort-wchar",
-        "-Wl,-nostdlib", "-T", ldscript, "-o", "zefi.elf", cfile]
+           "-I", include_dir, "-fno-stack-protector", "-fpic", "-mno-red-zone",
+           "-fshort-wchar", "-Wl,-nostdlib", "-T", ldscript, "-o", "zefi.elf", cfile]
```
Nothing wrong with this update, but a small proposal worth consider.

CMake is our build system, and is designed for handling compilation and linking.

So maybe it should be considered to change this script into generating source code, and then use that source code for an extra build target which generates the `zefi.elf`.

Some pros on such change:
- CMake can make sure `GENERATED` property is correctly set on the generated source.
- No need to also pass include folder(s) to the script.
- it is clear what happens during build, as the compilation and linking of `zefi.c` and `zefi.elf` will be part of `build.ninja` / `Makefile`. Thus also show up with `ninja -v`.
- The source file will be part of `compile_commands.json`, used by IDEs.
- zefi.elf will be dedicated target, meaning it will be possible to execute `ninja zefi.elf` / `make zefi,ekf` directly, and thus the target is built including its dependencies.

tbursztyka replied:
Looks like an enhancement to me. Definitely worth doing but not in this PR

tejlmand reviewed (dismissed):
Generally looks good.

One suggestion for this PR, and a proposal to consider for the future.

tejlmand reviewed (dismissed):
lgtm - re-approved

andyross reviewed (dismissed):
Refresh +1.  Interestingly I'm about to be diving back into EFI soon and this would be nice to have merged first.

tbursztyka commented:
> Refresh +1. Interestingly I'm about to be diving back into EFI soon and this would be nice to have merged first.

I'd like it too  
Unfortunately EHL issue is still eluding me, the data in zefi gets 0ed. I am missing something

Commit:
arch/x86: Initialize early serial a tiny bit later

In case of EFI, efi_init must be called before initializing early
serial: if that one as X86_SOC_EARLY_SERIAL_PCIDEV defined, its pcie
access will try to initialise pcie mmio access which one will try to
find an ACPI table. At this point, calling ACPI API prior to initialize
EFI will make RSDP looked up already... and since it cannot find it
without EFI being initialized first, ACPI is then broken.

Just moving early serial to initialize after multiboot/efi being setup.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

tbursztyka commented:
Spoke too fast, it was not 0ed after all. It was an acpi/efi ordering issue. (I trusted efi_get_rsdp_acpi() without thinking about such ordering...)

Now it works.

tejlmand commented:
lgtm - re-approved
